platform/default/src/mbgl/storage/pmtiles_file_source.cpp

```diff
@@ -0,0 +1,547 @@
+            if (directory_cache_control.at(url).back() != directory_cache_key) {
+                directory_cache_control.at(url).emplace_back(directory_cache_key);
+
+                for (auto it = directory_cache_control.at(url).begin(); it != directory_cache_control.at(url).end();
+                     ++it) {
+                    if (*it == directory_cache_key) {
+                        directory_cache_control.at(url).erase(it);
+                        break;
+                    }
+                }
+            }
+
+            callback(std::unique_ptr<Response::Error>());
+            return;
+        }
+
+        getHeaderAndRootDirectory(url, req, [=, this](std::unique_ptr<Response::Error> error) {
+            if (error) {
+                callback(std::move(error));
+                return;
+            }
+
+            pmtiles::headerv3 header = header_cache.at(url);
+
+            Resource resource(Resource::Kind::Source, url);
+            resource.loadingMethod = Resource::LoadingMethod::Network;
+            resource.dataRange = std::make_pair(directoryOffset, directoryOffset + directoryLength - 1);
+
+            tasks[req] = getFileSource()->request(resource, [=, this](const Response& response) {
+                if (response.error) {
+                    callback(std::make_unique<Response::Error>(
+                        response.error->reason,
+                        std::string("Error fetching PMTiles directory: ") + response.error->message));
+
+                    return;
+                }
+
+                try {
+                    std::string directoryData = *response.data;
+
+                    if (header.tile_compression == pmtiles::COMPRESSION_GZIP) {
+                        directoryData = util::decompress(directoryData);
+                    }
+
+                    storeDirectory(url, directoryOffset, directoryLength, directoryData);
+
+                    callback(std::unique_ptr<Response::Error>());
+                } catch (const std::exception& e) {
+                    callback(std::make_unique<Response::Error>(
+                        Response::Error::Reason::Other,
+                        std::string(std::string("Error parsing PMTiles directory: ") + e.what())));
+                }
+            });
+        });
+    }
+
+    void getTileAddress(const std::string& url,
+                        AsyncRequest* req,
+                        uint64_t tileID,
+                        uint64_t directoryOffset,
+                        uint32_t directoryLength,
+                        uint32_t directoryDepth,
+                        AsyncTileCallback callback) {
+        if (directoryDepth > 3) {
+            callback(std::make_pair<uint64_t, uint32_t>(0, 0),
+                     std::make_unique<Response::Error>(
+                         Response::Error::Reason::Other,
+                         std::string("Error fetching PMTiles tile address: Maximum directory depth exceeded")));
```

Comment:
I think you forgot to return here?

Reply:
You are right, I forgot! Thank you for pointing that.

Reply:
Done!